### ISSUE TITLE
Running systemctl reads from EFI variables when launched by a monitoring agent

### ISSUE THREAD

**Issue by filbranden** [systemd/systemd#16097]
**systemd version the issue has been seen with**
> v245-1586-gb11e98037c

**Used distribution**
 > Fedora 32

**Expected behaviour you didn't see**
 > systemctl shouldn't read EFI variables, at all.

**Unexpected behaviour you saw**
 > systemctl (and other tools, systemd-analyze, etc.) read EFI variables if they're called by a daemon, such as Chef agent, which runs them without a controlling tty.

**Steps to reproduce the problem**

This works fine:

```
$ strace -e openat ./build/systemctl 2>&1 | grep /sys/firmware
(empty)
```

But this doesn't:

```
$ strace -e openat setsid ./build/systemctl 2>&1 | grep /sys/firmware
openat(AT_FDCWD, "/sys/firmware/efi/efivars/SecureBoot-XXX-YYY-ZZZ", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/sys/firmware/efi/efivars/SystemdOptions-XXX-YYY-ZZZ", O_RDONLY|O_NOCTTY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

Ooops!

Looking at the source, I can see that `log_parse_environment()` tries to detect whether it's being run from a daemon by [checking whether there's a controlling tty](https://github.com/systemd/systemd/blob/v245/src/basic/log.c#L1112-L1116) when deciding to parse `/proc/cmdline`.

And recently commit https://github.com/systemd/systemd/commit/c7d26acce6dcb0e72be6160873fac758e9b7c440 (which was first released in v245) will read EFI variables when parsing `/proc/cmdline`, so this code got a lot more expensive, enough to be noticeable.

cc @anitazha @arianvp

**Comment by filbranden** [comment]
Since EFI changes the game... I wonder if it would make sense to have PID 1 parse `/proc/cmdline` only once and save the results in `/run/systemd/proc_cmdline` or similar and then have everyone else consume it from there?

Even better, we could pre-parse the logging configuration from the kernel command line and keep a pre-processed copy of that under `/run` so processes could already get the chewed up settings to use for logging...

**Comment by arianvp** [comment]
Also see the discussion in https://github.com/systemd/systemd/issues/14828#issuecomment-626167308

**Comment by poettering** [comment]
I figure #15977 also fixes this issue, no? 

@DaanDeMeyer I figure you can add `Fixes: #16097` to the commit message of #15977.

**Comment by filbranden** [comment]
> I figure #15977 also fixes this issue, no?

I just checked and it doesn't. Built a tree from that PR and I still see the access to the EFI variables.

```
$ strace -e openat setsid ./build/systemctl 2>&1 | grep /sys/firmware
openat(AT_FDCWD, "/sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/sys/firmware/efi/efivars/SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67", O_RDONLY|O_NOCTTY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

PR #16139 fixes the specific issue of command-line utilities such as `systemctl` *accessing EFI variables* when run from a daemon... But it doesn't address the more general issue of them still taking the kernel command-line (or EFI SystemdOptions) in consideration depending on whether they're attached to a console or not... Whether that's something that's important to fix (and that we want to fix) or not might be up for debate... PR #16098 (which is now closed) would address that part too. (Granted, it could use a lot more polishing.) I guess the big question for that one is whether we can cleanly split the commands into CLI only vs. Daemon only vs. Mixed-mode and have the last one of those be small enough (and/or tools that are used one way 99% of the time) that we're happy to do the console check on that one only.
